lib/dal/src/attribute/value.rs

```diff
@@ -2211,6 +2251,67 @@ impl AttributeValue {
         }
     }
 
+    ///
+    /// Get matching child attributes, in order.
+    /// 
+    pub async fn get_child_av_id_pairs_in_order(
+        ctx: &DalContext,
+        first_parent: AttributeValueId,
+        second_parent: AttributeValueId
+    ) -> AttributeValueResult<Vec<ChildAttributeValuePair>> {
+        let mut pair_index = HashMap::<KeyOrIndex, usize>::new();
+        let first_children = Self::get_child_av_ids_in_order(ctx, first_parent).await?;
+        let mut pairs: Vec<ChildAttributeValuePair> = Vec::with_capacity(first_children.len());
+        for (index, first_child) in first_children.into_iter().enumerate() {
+            let key = Self::key_for_id(ctx, first_child).await?;
+            let key_or_index = match &key {
+                Some(key) => KeyOrIndex::Key(key.clone()),
+                None => KeyOrIndex::Index(index as i64),
+            };
+            if let Some(old_index) = pair_index.get(&key_or_index) {
+                return Err(AttributeValueError::DuplicateKeyOrIndex {
+                    key_or_index,
+                    // It is impossible at this point to have anything except FirstOnly
+                    // in the vec.
+                    child1: pairs[*old_index].first().unwrap(),
```

Comment:
Although logically impossible to panic here, this should be written to be panic free, just in case:

`pairs.get(*old_index).and_then(|pair| pair.first())`. You can just make the child1 type on this error an `Option` to prevent having to `ok_or` or `unwrap`

Reply:
(our linter will not allow you to ship this with an unwrap, which is likely why it's failing)

Reply:
Do we mind the use of `unreachable!()` or will that trigger lints too?

Reply:
unreachable! won't trigger lints, no, and is fine if you know the spot is really unreachable. But if you can produce an error instead, that's always better

Reply:
I think I found a way to avoid the lint; as long as I keep first_children around I can just reach inside that to get the old ID.

Reply:
the direct indexing is also a panic case. if the index is not there, it panics

Reply:
Fixed as suggested by stuffing an existing index in there if it's None, and adding a comment noting that it should never happen.

Reply:
I just now grokked the putting an Option in the error. I'm ok either way if you'd prefer that!